### PR TITLE
Selector Signal fix with multiple workers

### PR DESCRIPTION
**Pull request by yuandrew**
## What was changed
Fixed a scenario where selectors without default branch was having shared data `c.recValue` overwritten. 

Added a new flag `SDKFlagWorkflowNewChannelLostMessages` to fix this, as there were 2 SDK versions (v1.35.0 and v1.36.0) that enabled `SDKFlagBlockedSelectorSignalReceive`, so we need a new flag so those histories don't break.

## Why?
Fix bug

Both flags will be enabled in a future release (both currently disabled, so no behavior change with this PR), after this new flag's introduction has baked at least for a release or 2. No point in re-enabling `SDKFlagBlockedSelectorSignalReceive` if we don't also enable `SDKFlagWorkflowNewChannelLostMessages`

## Checklist

1. #2066 and #1624 (will close issues when the SDK flag is enabled)

2. How was this tested:
Added new test

Added replay test verifying history with just `SDKFlagBlockedSelectorSignalReceive` doesn't break. This is primarily valuable once we enable `SDKFlagWorkflowNewChannelLostMessages` in a future release

---

> [!NOTE]
> **Medium Risk**
> Touches core deterministic workflow selector/channel internals; while gated behind a new default-off flag and covered by unit + replay tests, mistakes here could impact message delivery semantics or replay behavior.
> 
> **Overview**
> Fixes a workflow `Selector` receive edge case where, with `SDKFlagBlockedSelectorSignalReceive` enabled, multiple blocked selectors on the same channel could overwrite `channelImpl.recValue` and cause a dropped/zero-value receive.
> 
> Introduces `SDKFlagWorkflowNewChannelLostMessages` (default disabled) and updates `selectorImpl.Select` to only pre-store `recValue` in the callback when appropriate (e.g., when a default branch exists), otherwise deferring storage to the per-callback `readyBranch` to prevent cross-selector clobbering. Adds unit coverage for the “two workers on one channel” pattern and a replay test + history JSON to ensure old behavior remains replayable when only the older flag is present in history.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c618c78d3077849707df1cae502f409f3c083304. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: internal/internal_flags.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ const (
 
 // unblockSelectorSignal exists to allow us to configure the default behavior of
 // SDKFlagBlockedSelectorSignalReceive. This is primarily useful with tests.
-var unblockSelectorSignal = os.Getenv("UNBLOCK_SIGNAL_SELECTOR") != ""
```
Comment (Quinn-With-Two-Ns) [inline internal/internal_flags.go]: Is there any roll back concerns here? Do we need to protect this behaviour by another flag?

Comment (yuandrew) [reply internal/internal_flags.go]: Added another flag

File: test/replaytests/channel-worker-blocked-selector.json
Comment (cretz) [inline test/replaytests/channel-worker-blocked-selector.json:1]: Was this history generated from the current SDK before this flag and confirms with the new flag it still works?

I'm wondering if there are 3 SDK replay tests needed here: 1 from before a couple of versions ago, 1 from current released, and 1 from with this PR? If not no worries.

I admit I don't understand the PR or the problem well enough, just confirming coverage/confidence.

Comment (yuandrew) [reply test/replaytests/channel-worker-blocked-selector.json:1]: Yes, these were generated off of main before this solution (and flag) were introduced. 

Good question, the three scenarios you mentioned are these SDK flag combinations: none, 5, and 5+7.

`TestSelectorNonBlocking` already should cover `none`, previously implemented in the original solution, and this new test `TestChannelWorkerWithBlockedSelectorFlag` now covers flag 5. I'm not sure we need a replay test for 5+7, where we have unit tests that validate the behavior. 

I guess it wouldn't hurt to add, but I'm thinking replay tests are primarily used for ensuring old histories don't hit NDE when we introduce a new flag.

**Review by cretz — APPROVED** [review]
I admit I don't understand this very well and may not easily be able to, but if we have replay tests confirming all scenarios are safe, works for me.
